smart_campus/app/views.py

```diff
@@ -618,3 +639,390 @@ def reward_add_page(request):
+        travelplan_form = PartialTravelPlanForm(
+                                request.POST,
+                                request.FILES,
+                                instance=travelplan
+                            )
+        print(request.POST)
+        if travelplan_form.is_valid():
+            edited_travelplan = travelplan_form.save()
+
+            json_order = json.loads(request.POST['order'])
+
+            if not json_order:
+                for travelplan_station in TravelPlanStations.objects.filter(
+                                                travelplan_id=pk
+                                            ):
+                    travelplan_station.delete()
+
+            else:
+                for order, station_id in enumerate(json_order):
+                    changed_travelplan = TravelPlanStations.objects.filter(
+                                                travelplan_id=pk,
+                                                station_id=station_id
+                                            )
+
+                    if not changed_travelplan:
+                        TravelPlanStations.objects.create(
+                            travelplan_id=pk,
+                            station_id=station_id,
+                            order=order
+                        )
+                    else:
+                        changed_travelplan.first().order = order
+                        changed_travelplan.first().save()
+
+            context = {
+                'categories': StationCategory.objects.all(),
+                'email': request.user.email,
+                'travelplans': TravelPlan.objects.all()
+            }
+
+            return HttpResponseRedirect('/travelplans/')
+
+        form_data = travelplan_form.cleaned_data
+
+    else:
+        form_data = {
+            'name': travelplan.name,
+            'description': travelplan.description
+        }
+
+    travelplanstations = TravelPlanStations.objects.filter(
+                                travelplan_id=pk
+                            ).order_by('order')
+
+    selected_stations_id = [
+        travelplanstation.station_id for travelplanstation
+        in travelplanstations
+    ]
+
+    selected_stations = [
+        Station.objects.get(id=station_id) for station_id
+        in selected_stations_id
+    ]
+
+    context = {
+        'email': request.user.email,
+        'stations': Station.objects.all(),
+        'travelplan': travelplan,
+        'travelplanstations': travelplanstations,
+        'selected_stations': selected_stations,
+        'form_data': form_data,
+        'categories': StationCategory.objects.all()
+    }
+
+    return render(request, 'app/travelplan_edit_page.html', context)
+
+
+@login_required
+def travelplan_delete_page(request, pk):
+    # Delete the travelplan and its travelplan_stations
+
+    travelplan = get_object_or_404(TravelPlan, pk=pk)
+    travelplan_stations = TravelPlanStations.objects.filter(travelplan_id=pk)
```

Comment:
Why don't you set "on delete cascade" in the model?

Reply:
Did you mean this below?
```
    stations = models.ManyToManyField(
        'Station',
        through='TravelPlanStations'
    )
```
I have no Idea about that. Just follow the model.

Reply:
I thought it was due to "Django didn't have ON DELETE CASCADE option for manytomanyfield", I didn't find out any of this mentioned in Django documentation.

ref1: [stackoverflow discuss](https://stackoverflow.com/questions/3937194/django-cascade-deletion-in-manytomanyrelation)
* I thought the discuss above making it very reasonable.

ref2: [Django Documentation about manytomanyfield](https://docs.djangoproject.com/en/1.11/topics/db/examples/many_to_many/)
ref3: [Django Documentation about Model field](https://docs.djangoproject.com/en/1.11/ref/models/fields/)

Reply:
@Lee-W

Reply:
@Lee-W